Data Explorer: Use complete URIs rather than file paths for DuckDB

atheriel opened the pull request:
This commit refactors the DuckDB data explorer backend and frontend to operate on complete URIs rather than filepaths, part of enabling the data explorer to support virtual filesystem providers. It almost all comes down to swapping out a series of string-typed fields for `vscode.Uri`-typed fields.

~~In local testing, I've been able to use this to open files from custom filesystem providers in the data explorer.~~

Part of #7351.

### Release Notes

#### New Features

- The data explorer can now open files backed by virtual filesystem providers.

#### Bug Fixes

- N/A

### QA Notes

It's not easy to test against virtual filesystem providers, unfortunately. [VS Code suggests using the GitHub Repositories extension](https://code.visualstudio.com/api/extension-guides/virtual-workspaces), but that's not available in OpenVSX if I understand correctly.

atheriel committed:
Data Explorer: Use complete URIs rather than file paths for DuckDB.

This commit refactors the DuckDB data explorer backend and frontend to
operate on complete URIs rather than filepaths, the first step in
enabling the data explorer to support virtual filesystem providers. It
almost all comes down to swapping out a series of string-typed fields
for vscode.Uri-typed fields.

Since the generated RPC definitions can't currently handle URIs, we
roundtrip through toString() and Uri.parse() at the comm layer.

Note that we still rely on `fs` in the extension, which doesn't work
with non-file URIs, so I added an error message that should surface to
users attempting to open these unsupported URIs. Previously the data
explorer would just be empty when attempting this.

Part of #7351.

Signed-off-by: Aaron Jacobs <aaron.jacobs@posit.co>

wesm commented on extensions/positron-duckdb/src/interfaces.ts:
```diff
@@ -1252,7 +1254,7 @@ export interface OpenDatasetParams {
 	 */
-	uri: string;
+	uri: vscode.Uri;
 }
```
This is tricky — since this uri parameter is defined in the data explorer backend RPC definition, probably instead what we want to do is define a URI data structure there that can be cast back and forth to a `vscode.Uri`. I can take a closer look at this

atheriel replied:
Maybe we can just leave it as a string and roundtrip the URI structure via `uri.toString()` -> `vscode.Uri.parse()`.

atheriel replied:
Updated the PR to do exactly this, which should leave the generated files unchanged.

wesm commented on src/vs/workbench/services/languageRuntime/common/positronDataExplorerComm.ts:
```diff
@@ -1429,7 +1430,7 @@ export class PositronDataExplorerComm extends PositronBaseComm {
-	openDataset(uri: string): Promise<OpenDatasetResult> {
+	openDataset(uri: URI): Promise<OpenDatasetResult> {
 		return super.performRpc('open_dataset', ['uri'], [uri]);
 	}
```
This code is generated so as I mentioned above we'll have to figure out an alternate approach that is compatible with the code generator. Shouldn't be too hard.

wesm reviewed (commented):
Thanks for doing this. We just need to resolve the sharp edge with the OpenRPC code generator to make this work. I can take over that piece and help get this over the finish line

wesm reviewed (commented):
These code changes look OK now with the code generator issue out of the way. The UI isn't letting me click the merge button so I'm not sure how to proceed, do you want to just push these changes to the other PR?

atheriel commented:
I've updated this to merge directly against `main` and deferred the `fs`-related changes. So this commit enables URIs everywhere, but doesn't support non-`file` schemes.

I added an error screen for users that explains this; previously the data explorer would just be blank in these cases.

![Screenshot from 2025-04-30 20-01-18](https://github.com/user-attachments/assets/6cf23194-907d-4f21-84f6-c3e692baec8f)

wesm reviewed (approved):
+1 -- was finally able to test this out on Windows (make sure that the file watching / live update logic continues to work) and all looks good. Thanks for doing this!

atheriel commented:
@jonvanausdeln Ah, I see why; I forgot about `vscode-remote://` URLs being valid. This should be fixed by #7352, but I can also add a more targeted fix if there are other issues with that PR. What do you think @wesm?

wesm commented:
That PR looks OK to me and all the e2e data-explorer tests were run there (didn't enable them here with @:data-explorer, my bad!) so I will merge that now to see if the fix works.
